src/main/java/com/rpi/alexandria/service/UserService.java

```diff
@@ -82,6 +81,54 @@ public UserDetails loadUserByUsername(final String username) throws UsernameNotF
 		return toUserDetails(user);
 	}
 
+	public boolean hasAddedRating(User loggedInUser, Rating rating) {
+		return loggedInUser.hasRated(rating.getUsernameOther());
+	}
+
+	public void addRating(User loggedInUser, Rating rating) {
+		String usernameOther = rating.getUsernameOther();
+
+		User otherUser = getUser(usernameOther);
+		otherUser.addRating(loggedInUser.getUsername(), rating.getRatingValue());
+
+		loggedInUser.addUsernameToUsernamesRated(usernameOther);
+
+		userRepository.save(otherUser);
+		userRepository.save(loggedInUser);
+	}
+
+	public void updateRating(String loggedInUserUsername, Rating rating) {
+		User otherUser = getUser(rating.getUsernameOther());
+		otherUser.updateRating(loggedInUserUsername, rating.getRatingValue());
+		userRepository.save(otherUser);
+	}
+
+	public double getAverageRating(Rating rating) {
+		User user = getUser(rating.getUsernameOther());
+		return user.getAverageRating();
+	}
+
+	public Map<String, Integer> getRatings(Rating rating) {
+		User user = getUser(rating.getUsernameOther());
+		return user.getRatings();
+	}
+
+	public Set<String> getUsernamesRated(User user) {
+		return user.getUsernamesRated();
+	}
+
+	public void deleteRating(User loggedInUser, Rating rating) {
+		String usernameOther = rating.getUsernameOther();
+
+		User otherUser = getUser(usernameOther);
+		otherUser.deleteRating(loggedInUser.getUsername());
+
+		loggedInUser.removeUsernameFromUsernamesRated(usernameOther);
+
+		userRepository.save(otherUser);
+		userRepository.save(loggedInUser);
```

Comment:
change this to `userRepository.saveAll(List.of(otherUser, loggedInUser));`

Reply:
updated